dbms/src/Encryption/AESCTRCipher.h

```diff
@@ -0,0 +1,87 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <Common/config.h>
+#include <common/types.h>
+#include <openssl/aes.h>
+#include <openssl/conf.h>
+#if USE_GM_SSL
+#include <gmssl/sm4.h>
+#endif
+#include <Storages/KVStore/FFI/FileEncryption.h>
+
+#if OPENSSL_VERSION_NUMBER < 0x01010000f
+
+#define InitCipherContext(ctx) \
+    EVP_CIPHER_CTX ctx##_var;  \
+    ctx = &ctx##_var;          \
+    EVP_CIPHER_CTX_init(ctx);
+
+// do nothing
+#define FreeCipherContext(ctx)
+
+#else
+
+#define InitCipherContext(ctx)              \
+    ctx = EVP_CIPHER_CTX_new();             \
+    if ((ctx) != nullptr)                   \
+    {                                       \
+        if (EVP_CIPHER_CTX_reset(ctx) != 1) \
+        {                                   \
+            (ctx) = nullptr;                \
+        }                                   \
+    }
+
+#define FreeCipherContext(ctx) EVP_CIPHER_CTX_free(ctx);
+
+#endif
+
+#if (USE_GM_SSL == 0) && !defined(OPENSSL_NO_SM4)
+// TODO: OpenSSL Lib does not export SM4_BLOCK_SIZE by now.
+// Need to remove SM4_BLOCK_SIZE once Openssl lib support the definition.
+// SM4 uses 128-bit block size as AES.
+// Ref:
+// https://github.com/openssl/openssl/blob/OpenSSL_1_1_1-stable/include/crypto/sm4.h#L24
+#define SM4_BLOCK_SIZE 16
+#endif
```

Comment:
I think we can move these into the `.cpp` file

Reply:
done